packages/in_app_purchase/android/src/main/java/io/flutter/plugins/inapppurchase/InAppPurchasePlugin.java

```diff
@@ -50,8 +50,11 @@
   public static void registerWith(Registrar registrar) {
     InAppPurchasePlugin plugin = new InAppPurchasePlugin();
     plugin.setupMethodChannel(registrar.activity(), registrar.messenger(), registrar.context());
-    ((Application) registrar.context())
-        .registerActivityLifecycleCallbacks(plugin.methodCallHandler);
+    if (registrar.context().getApplicationContext() != null
+        && registrar.context().getApplicationContext() instanceof Application) {
```

Comment:
Is it bad if it's null or not an Application?  Does that need to be handled?

Reply:
if either of that situation happens, we are not able to get the application, which makes us no way to register life cycle callbacks. We kind of just bailing here.
I'm not sure if there would be a way to handle that situation in V1 embedding, but we definitely don't want to crash here.

Reply:
Do we know how people are getting in this state?

Reply:
I am not entirely sure. @mklim or @matthew-carroll might have more ideas on this.

Reply:
I believe that you can just say `registrar.context().getApplicationContext()`. I don't think that `context()` can return null, and I don't think that `getApplicationContext()` can return null

Reply:
@matthew-carroll Thanks, will getApplicationContext always be able to cast to Application?

Reply:
I believe so. As long as the `context` on which you're invoking `getApplicationContext()` is non-null, I think the line will always be fine.

Reply:
BTW, do you know why you need an `Application`? What are the `Application` methods that are being called that are not defined on a `Context`?

Reply:
ok. I just did some search and it seems to be the case. I'll update the fix.

Reply:
I believe `.registerActivityLifecycleCallbacks` is only in `Application`?

Reply:
That looks right.

Reply:
To explain this a little more, there is a difference between an `ApplicationContext` (can be cast to `Application`) and an `ActivityContext` (can be cast to `Activity`) even though they're both `Context`s. Generally it's hard to tell the difference but occasionally it's critical. The Activity Context is tied to the Activity lifecycle so can be invalidated much easier.